xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java

```diff
@@ -0,0 +1,479 @@
+        nonEmptySince = timeProvider.currentTimeNanos();
+      }
+      lastUpdated = timeProvider.currentTimeNanos();
+      weight = newWeight;
+    }
+
+    @Override
+    public void start(SubchannelStateListener listener) {
+      delegate().start(new SubchannelStateListener() {
+        @Override
+        public void onSubchannelState(ConnectivityStateInfo newState) {
+          if (newState.getState().equals(ConnectivityState.READY)) {
+            nonEmptySince = Integer.MAX_VALUE;
+          }
+          listener.onSubchannelState(newState);
+        }
+      });
+    }
+
+    private double getWeight() {
+      if (config == null) {
+        return 0;
+      }
+      double now = timeProvider.currentTimeNanos();
+      if (now - lastUpdated >= config.weightExpirationPeriodNanos) {
+        nonEmptySince = Integer.MAX_VALUE;
+        return 0;
+      } else if (now - nonEmptySince < config.blackoutPeriodNanos) {
+        return 0;
+      } else {
+        return weight;
+      }
+    }
+
+    @Override
+    protected Subchannel delegate() {
+      return delegate;
+    }
+  }
+
+  @VisibleForTesting
+  final class WeightedRoundRobinPicker extends ReadyPicker {
+    private final List<Subchannel> list;
+    private final AtomicReference<EdfScheduler> schedulerRef;
+    private volatile boolean rrMode = false;
+
+    WeightedRoundRobinPicker(List<Subchannel> list, int startIndex) {
+      super(list, startIndex);
+      Preconditions.checkArgument(!list.isEmpty(), "empty list");
+      this.list = list;
+      this.schedulerRef = new AtomicReference<>(new EdfScheduler());
+      updateWeight();
+    }
+
+    @Override
+    public PickResult pickSubchannel(PickSubchannelArgs args) {
+      if (rrMode) {
+        return super.pickSubchannel(args);
+      }
+      int pickIndex = schedulerRef.get().pick();
+      WrrSubchannel subchannel = (WrrSubchannel) list.get(pickIndex);
+      if (!config.enableOobLoadReport) {
+        return PickResult.withSubchannel(
+           subchannel,
+           OrcaPerRequestUtil.getInstance().newOrcaClientStreamTracerFactory(
+               subchannel.perRpcListener));
+      } else {
+        return PickResult.withSubchannel(subchannel);
+      }
+    }
+
+    private void updateWeight() {
+      EdfScheduler scheduler = new EdfScheduler();
+      int weightedChannelCount = 0;
+      double avgWeight = 0;
+      for (Subchannel value : list) {
+        double newWeight = ((WrrSubchannel) value).getWeight();
+        if (newWeight > 0) {
+          avgWeight += newWeight;
+          weightedChannelCount++;
+        }
+      }
+      rrMode = weightedChannelCount < 2;
+      if (rrMode) {
+        return;
+      }
+      for (int i = 0; i < list.size(); i++) {
+        WrrSubchannel subchannel = (WrrSubchannel) list.get(i);
+        double newWeight = subchannel.getWeight();
+        scheduler.add(i, newWeight > 0 ? newWeight : avgWeight);
+      }
+      schedulerRef.set(scheduler);
```

Comment:
Absolutely true.
Note stubby has similar "completion ratio" mechanism, that gives credits to subchannels in the previous state when updating to the next state with the new weight. This way, the minimum weight channel can possibly be picked.
The current implementation is very simplified. I'll make it as a future improvement and I will capture it in the design doc.

Reply:
> Because you are completely replacing the scheduler each time you update weights, if something has a large weight and something else has a minimum weight the second one might never get used.

Randomizing the scheduler each creation should prevent that. Worst-case, if a schedule is used for only a single pick after being created, then that is the same as a WRR implementation that has weighted ranges for each choice and uses a random number to choose (same approach as weighted_target).

The code right now does not seem to randomize the initial scheduler state, but it will need to before de-experimentalizing.